### PR TITLE
Add initialization for coordinator

### PR DESCRIPTION
**Pull request by gfphoenix78**
## Description

Initialize coordinator, includes:
* init shared memory, (memory for storing inspecting data is not ready now)
* starts two background workers, one is for the coordinator and the other is for auxiliary process
* init UNIX domain socket for the coordinator to listen to messages from QE/plcontainer
* init loop framework

Co-authored-by: Xinyue Ge <xge@pivotal.io>
Co-authored-by: Hao Wu <hawu@pivotal.io>

### CODE REVIEW

File: src/co_coordinator.c [hunk 1]
```diff
@@ -33,108 +33,165 @@
...
 
-void
-plc_coordinator_main(Datum datum)
+static int
+plc_listen_socket()
 {
-    // TODO: impl coordinator logic here
+    char address[500];
+    int sock;
+    snprintf(address, sizeof(address), "/tmp/plcoordinator_%d_unix.sock", (int)getpid());
```
Comment (HaozhouWang) [inline src/co_coordinator.c]: It is better to change to .plcoordinator.%d.unix.sock and hidden from the user.

File: src/co_coordinator.c [hunk 2]
```diff
@@ -33,108 +33,165 @@
...
+        /* TODO: add network code */
+
+        rc = WaitLatch(&MyProc->procLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH, 0);
+        if (rc & WL_POSTMASTER_DEATH)
+            break;
+        ResetLatch(&MyProc->procLatch);
+        if (got_sighup) {
+            plc_refresh_container_config(false);
+        }
+        sleep(2);
```
Comment (HaozhouWang) [inline src/co_coordinator.c]: It is better to add a comment here why we sleep 2 seconds?

File: src/co_coordinator.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#ifndef _CO_COORDINATOR_H
```
Comment (HaozhouWang) [inline src/co_coordinator.h]: Add license header

File: src/co_coordinator.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    coordinator_protocol protocol;
+    char address[504];
+} CoordinatorStruct;
+#define CO_SHM_KEY  "plcoordinator_shm"
+
+typedef struct
+{
+    int id; // pid of QE
+    int sock; // socket file descriptor between QE and coordinator
+} requester_info_entry;
```
Comment (HaozhouWang) [inline src/co_coordinator.h]: Do we need to store the id-sock pair in memory? If yes, why? It seems that the connection will be closed after the coordinator sends results back to GPDB and will not be reused anymore.

**Comment by HaozhouWang** [comment]
It is better to move coordinator code into a single folder.

File: src/co_coordinator.c (hunk 2 above)
Comment (gfphoenix78) [reply src/co_coordinator.c]: It's a dummy line here. To make the function not be a busy loop

File: src/co_coordinator.c (hunk 1 above)
Comment (gfphoenix78) [reply src/co_coordinator.c]: OK, thanks!

File: src/co_coordinator.h (hunk 3 above)
Comment (gfphoenix78) [reply src/co_coordinator.h]: Thank you!

File: src/co_coordinator.h (hunk 4 above)
Comment (gfphoenix78) [reply src/co_coordinator.h]: This struct entry in a WIP status. The connection may be or be not closed after the first allocation of container. Maybe there is some other information to add that can't figure out now. In the other hand, if the socket file descriptor is needless, the whole hash map can also be removed.
The next step will determine whether to keep or remove it.

Comment (HaozhouWang) [reply src/co_coordinator.h]: Sounds good!

File: src/co_coordinator.c (hunk 2 above)
Comment (HaozhouWang) [reply src/co_coordinator.c]: Got it!

**Pull request merged**